### PR TITLE
[BIOMAGE-1670] - Dispatch and receive GetExpressionCellSets schema

### PR DESCRIPTION
**Pull request by aerlaut**
# Description
- Dispatch work request for GetExpressionCellSets
- Receive work request and create cell set
- Setup loading states

# Details
#### URL to issue
https://biomage.atlassian.net/browse/BIOMAGE-1670

#### Link to staging deployment URL (or set N/A)
https://ui-agi-ui630-api297.scp-staging.biomage.net/

#### Links to any PRs or resources related to this PR
https://github.com/hms-dbmi-cellenics/ui/pull/630
https://github.com/hms-dbmi-cellenics/api/pull/297
https://github.com/hms-dbmi-cellenics/worker/pull/240

#### Integration test branch
master

# Merge checklist
Your changes will be ready for merging after all of the steps below have been completed.

### Code updates
Have best practices and ongoing refactors being observed in this PR
- [X] Migrated any selector / reducer used to the new format.

### Manual/unit testing
- [X] Tested changes using InfraMock locally **or** no tests required for change, e.g. Kubernetes chart updates.
- [X] Validated that current unit tests for code work as expected and are sufficient for code coverage **or** no unit tests required for change, e.g. documentation update.
- [X] Unit tests written **or** no unit tests required for change, e.g. documentation update.

### Integration testing
**You must check the box below** to run integration tests on the latest commit on your PR branch.
Integration tests have to pass before the PR can be merged. Without checking the box, your PR
**will not pass** the required status checks for merging.

- [x] Started end-to-end tests on the latest commit.

### Documentation updates
- [X] Relevant Github READMEs updated **or** no GitHub README updates required.
- [X] Relevant Wiki pages created/updated **or** no Wiki updates required.

### Optional
- [ ] Staging environment is unstaged before merging.
- [ ] Photo of a cute animal attached to this PR.

### CODE REVIEW

File: src/components/data-exploration/generic-gene-table/ExpressionCellSetModal.jsx [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+import {
+  Modal, Form, Button, Select, InputNumber, Row, Alert, Col,
+} from 'antd';
+
+import createCellSetByExpression from 'redux/actions/cellSets/createCellSetByExpression';
+
+const formItemStyle = { margin: '0.375rem 0' };
+const FORM_NAME = 'filterForm';
+
+const ExpressionCellSetModal = (props) => {
```
Comment (ivababukova) [inline src/components/data-exploration/generic-gene-table/ExpressionCellSetModal.jsx:14]: I would call it `CellSetByExpressionModal`

**Comment by ivababukova** [comment]
Agi, can you fix this console warning that appears when I click on `cellSet` button:
Screenshot 2022-02-11 at 09 10 49
?

File: src/utils/experimentUpdatesHandler.js [hunk 2]
```diff
@@ -68,12 +70,18 @@ const onGEM2SUpdate = (update, dispatch, experimentId) => {
 };
 
 const onWorkResponseUpdate = (update, dispatch, experimentId) => {
-  const { request, response: { error } } = update;
+  const { request: { body }, response: { error } } = update;
 
-  if (request.body.name === 'ClusterCells') {
-    if (error) throw new Error(error);
+  if (error) throw new Error(error);
```
Comment (ivababukova) [inline src/utils/experimentUpdatesHandler.js:75]: what was happening in the past when we weren't throwing an error for all types of body.name?

File: src/redux/actions/cellSets/createCellSetByExpression.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  };
+
+  try {
+    const result = await fetchWork(experimentId, body, getState, { broadcast: true });
+    return result;
+  } catch (e) {
+    console.error(e);
+    pushNotificationMessage('error', endUserMessages.ERROR_FETCHING_CELL_SETS);
+    throw new Error(e);
+  }
```
Comment (ivababukova) [inline src/redux/actions/cellSets/createCellSetByExpression.js:20]: Where are the actions types for this? `CELL_SET_BY_EXPRESSION_CREATING`,  `CELL_SET_BY_EXPRESSION_CREATED`,  `CELL_SET_BY_EXPRESSION_ERROR`?

File: src/utils/experimentUpdatesHandler.js [hunk 4]
```diff
@@ -68,12 +70,18 @@ const onGEM2SUpdate = (update, dispatch, experimentId) => {
...
 
+  if (body.name === 'ClusterCells') {
     dispatch(updateCellSetsClustering(experimentId));
+    pushNotificationMessage('success', endUserMessages.SUCCESS_CELL_SETS_RECLUSTERED);
+  }
+
+  if (body.name === 'GetExpressionCellSets') {
+    dispatch(loadCellSets(experimentId, true));
+    pushNotificationMessage('success', endUserMessages.SUCCESS_NEW_CLUSTER_CREATED);
   }
```
Comment (ivababukova) [inline src/utils/experimentUpdatesHandler.js]: I thought that the idea is to call `loadCellSets` from within the individual action for creating cell sets based on gene expression, called `createCellSetByExpression`, if the worker-results bucket contains results

File: src/utils/experimentUpdatesHandler.js (hunk 2 above)
Comment (aerlaut) [reply src/utils/experimentUpdatesHandler.js:75]: Previously there was only 1 work request which uses this path, `ClusterCells`, so the error is handled inside an if statement which checks if the body of the work result is `ClusterCells` before checking the error. Before that, we did not use the `ExperimentUpdates-` channel for broadcasting completion of work request results.

File: src/utils/experimentUpdatesHandler.js (hunk 4 above)
Comment (aerlaut) [reply src/utils/experimentUpdatesHandler.js]: We'll still insert the work result into the `worker-results` bucket, but we'll not depend on its content to trigger `loadCellSet` or not. If we depend on the content, only the client (user's browser) which triggers the work request will receive the work result and reload cellset. Other clients looking at the same experiment will not know that a new cell set has been created.

I think the trigger to tell the UI to reload cell sets should depend on the message returned from the API, which broadcasts the result of the worker result. The UI can check the broadcast message. If it contains an error, it should not reload the cellsets.

File: src/utils/experimentUpdatesHandler.js (hunk 2 above)
Comment (aerlaut) [reply src/utils/experimentUpdatesHandler.js:75]: In this implementation, all types of messages returned from the API will be able to trigger an error. If we handle the error after we check the body name, we'd be able to check if there returned message is correct (has a valid body name) before acting on the error.

**Comment by aerlaut** [comment]
> Agi, can you fix this console warning that appears when I click on `cellSet` button: Screenshot 2022-02-11 at 09 10 49 ?

fixed in [1bdcbf6](https://github.com/hms-dbmi-cellenics/ui/pull/630/commits/1bdcbf65333b1619b2b47a3d3c8c500c722a9345)

**Review by ivababukova — COMMENTED** [review]
Code looks good, but before I approve I want to test it end to end in staging. Can you tell me once it's ready?

File: src/__test__/components/data-exploration/generic-gene-table/ExpressionCellSetModal.test.jsx [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+const loadingText = /Creating cell set.../i;
+
+describe('ExpressionCellSetModal', () => {
+  beforeEach(async () => {
+    jest.clearAllMocks();
+
+    storeState = makeStore();
+
+    fetchMock.resetMocks();
+    fetchMock.mockIf(/.*/, mockAPI(generateDefaultMockAPIResponses(experimentId)));
```
Comment (kafkasl) [inline src/__test__/components/data-exploration/generic-gene-table/ExpressionCellSetModal.test.jsx:53]: what does thi `mockIf` do?

Comment (aerlaut) [reply src/__test__/components/data-exploration/generic-gene-table/ExpressionCellSetModal.test.jsx:53]: This setsup mock for a specific URL. If the requested URL matches the pattern, it'll be passed to the callback function. Our callback function handles all routes, so it's matched against all URLs.

**Review by ivababukova — APPROVED** [review]

**Pull request merged**